### PR TITLE
feat: add a simple scan table method with filter option

### PR DESCRIPTION
**Pull request by jkdowdle**
## Motivation

Found a need for using a scan operation on a dynamo table and I wanted to use `dynamost`

If we prefer to not support filtering quite yet (could be some edge cases I've not considered) we could consider just doing https://github.com/lifeomic/dynamost/pull/16 for now

### CODE REVIEW

File: src/dynamo-table.ts [hunk 1]
```diff
@@ -368,6 +389,44 @@ export class DynamoTable<
...
+    };
+  }
+
+  /**
+   * Scans the database
+   */
+  async scan(
+    options: ScanOptions<z.infer<Schema>> = {},
+  ): Promise<ScanResponse<z.infer<Schema>>> {
+    return this._scan<z.infer<Schema>>(options);
```
Comment (aecorredor) [inline src/dynamo-table.ts:427]: Is there any reason for having `_scan` and not just having the logic directly live here?

File: src/dynamo-table.ts [hunk 2]
```diff
@@ -368,6 +389,44 @@ export class DynamoTable<
     return this._query({ index, key, options });
   }
 
+  private async _scan<Entity>(options: ScanOptions<Entity>) {
+    const {
+      ConditionExpression,
+      ExpressionAttributeNames,
+      ExpressionAttributeValues,
+    } = options.filter
+      ? serializeCondition(options.filter)
```
Comment (aecorredor) [inline src/dynamo-table.ts:398]: nit: couldn't we just spread this directly when calling scan and avoid the `undefined` defaults?

**Review by aecorredor — COMMENTED** [review]
What are the filtering edge cases you were thinking of? I think this is better than #16.

File: src/dynamo-table.ts (hunk 1 above)
Comment (swain) [reply src/dynamo-table.ts:427]: +1

**Review by swain — DISMISSED** [review]
+1 to Alejandro's comment about dropping `_scan`. Otherwise, this looks great!

File: src/dynamo-table.test.ts [hunk 3]
```diff
@@ -250,6 +250,22 @@ describe('DynamoTable', () => {
           createdAt: '2020-01-01T23:00:00.000Z',
         },
       ]);
+
+      const usersByMultipleConditions = await userTable.scan({
+        filter: {
+          'greater-than': { id: 'user-1' },
+          'less-than': { id: 'user-3' },
+        },
+      });
```
Comment (swain) [inline src/dynamo-table.test.ts:259]: Can you add a test that scans _without_ a filter?

Comment (jkdowdle) [reply src/dynamo-table.test.ts:259]: @swain do you mean also without a limit? https://github.com/lifeomic/dynamost/pull/17/files#diff-78b2a4fa0ddcdec4cd4b8f43d4cd1f2a27d3afa4f277f030eb464d89f386b421R163-R165

as in no params?

**Review by aecorredor — APPROVED** [review]

**Review by aecorredor — APPROVED** [review]

**Pull request merged**